### ISSUE TITLE
Fasta crash: Unhandled FunctionNode

### ISSUE THREAD

**Issue by lukechurch** [dart-lang/sdk#29979]
@peter-ahe-google
ID: 33afd1be9008543d1cb54d750d3a830428af2537
```
Unhandled exception:
Unhandled: FunctionNode
#0      internalError (package:front_end/src/fasta/errors.dart:38:5)
#1      BodyBuilder.toValue (package:front_end/src/fasta/kernel/body_builder.dart:226:14)
#2      BodyBuilder.popForValue (package:front_end/src/fasta/kernel/body_builder.dart:207:31)
#3      BodyBuilder.handleParenthesizedExpression (package:front_end/src/fasta/kernel/body_builder.dart:734:44)
#4      Parser.parseParenthesizedExpression (package:front_end/src/fasta/parser/parser.dart:3062:14)
#5      Parser.parseParenthesizedExpressionOrFunctionLiteral (package:front_end/src/fasta/parser/parser.dart:3045:15)
#6      Parser.parsePrimary (package:front_end/src/fasta/parser/parser.dart:3008:14)
#7      Parser.parseUnaryExpression (package:front_end/src/fasta/parser/parser.dart:2942:14)
#8      Parser.parsePrecedenceExpression (package:front_end/src/fasta/parser/parser.dart:2811:13)
#9      Parser.parseExpression (package:front_end/src/fasta/parser/parser.dart:2784:11)
#10     Parser.parseExpressionStatement (package:front_end/src/fasta/parser/parser.dart:2700:13)
#11     Parser.parseStatementX (package:front_end/src/fasta/parser/parser.dart:2501:14)
#12     Parser.parseStatement (package:front_end/src/fasta/parser/parser.dart:2433:20)
#13     Parser.parseFunctionBody (package:front_end/src/fasta/parser/parser.dart:2363:15)
#14     DietListener.parseFunctionBody (package:front_end/src/fasta/source/diet_listener.dart:545:14)
#15     DietListener.buildFunctionBody (package:front_end/src/fasta/source/diet_listener.dart:433:5)
#16     DietListener.endTopLevelMethod (package:front_end/src/fasta/source/diet_listener.dart:214:5)
#17     Parser.parseTopLevelMethod (package:front_end/src/fasta/parser/parser.dart:1427:14)
#18     Parser.parseTopLevelMember (package:front_end/src/fasta/parser/parser.dart:1333:11)
#19     Parser._parseTopLevelDeclaration (package:front_end/src/fasta/parser/parser.dart:372:14)
#20     Parser.parseTopLevelDeclaration (package:front_end/src/fasta/parser/parser.dart:344:13)
#21     Parser.parseUnit (package:front_end/src/fasta/parser/parser.dart:336:15)
#22     SourceLoader.buildBody (package:front_end/src/fasta/source/source_loader.dart:146:14)
<asynchronous suspension>
#23     Loader.buildBodies (package:front_end/src/fasta/loader.dart:117:13)
<asynchronous suspension>
#24     KernelTarget.buildProgram (package:front_end/src/fasta/kernel/kernel_target.dart:282:20)
<asynchronous suspension>
#25     CompileTask.compile (package:front_end/src/fasta/fasta.dart:164:38)
<asynchronous suspension>
#26     compile.<anonymous closure> (package:front_end/src/fasta/fasta.dart:101:25)
<asynchronous suspension>
#27     CompilerContext.withGlobalOptions.<anonymous closure> (package:front_end/src/fasta/compiler_context.dart:45:33)
#28     _rootRun (dart:async/zone.dart:1120)
#29     _CustomZone.run (dart:async/zone.dart:1001)
#30     runZoned (dart:async/zone.dart:1467)
#31     CompilerContext.withGlobalOptions (package:front_end/src/fasta/compiler_context.dart:45:12)
#32     CompilerCommandLine.withGlobalOptions (package:front_end/src/fasta/compiler_command_line.dart:120:28)
#33     compile (package:front_end/src/fasta/fasta.dart:94:38)
<asynchronous suspension>
#34     compileEntryPoint (package:front_end/src/fasta/fasta.dart:53:11)
<asynchronous suspension>
#35     main (file:///home/lukechurch/GitRepos/dart-sdk/sdk/pkg/front_end/tool/_fasta/compile.dart:7:33)
#36     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:263)
#37     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)

```

**Comment by peter-ahe-google** [comment]
Repro:

```
main() {
  (f() {})();
}
```

**Comment by peter-ahe-google** [comment]
Proposed fix in [CL 2968093003](https://codereview.chromium.org/2968093003).
